src/parser/Parser.cpp

```diff
@@ -111,6 +143,153 @@ std::unique_ptr<AST::BlockStmt> Parser::parseBlockStmt() {
+    return expr;
+}
+
+std::unique_ptr<AST::Expr> Parser::parseEqExpr() {
+    std::unique_ptr<AST::Expr> expr = parseCompExpr();
+    while (m_currToken.type == TokenType::EqualEqual || m_currToken.type == TokenType::BangEqual) {
+        TokenType op = m_currToken.type;
+        eatToken(m_currToken.type);
+        auto right = parseCompExpr();
+        expr = std::unique_ptr<AST::BinaryExpr>(new AST::BinaryExpr(op,
+                                                                    std::move(expr),
+                                                                    std::move(right)));
+    }
+    return expr;
+}
+
+std::unique_ptr<AST::Expr> Parser::parseCompExpr() {
+    std::unique_ptr<AST::Expr> expr = parseConcatExpr();
+    while (m_currToken.type == TokenType::GreaterEqual ||
+           m_currToken.type == TokenType::LesserEqual ||
+           m_currToken.type == TokenType::Greater ||
+           m_currToken.type == TokenType::Lesser) {
+        TokenType op = m_currToken.type;
+        eatToken(m_currToken.type);
+        auto right = parseConcatExpr();
+        expr = std::unique_ptr<AST::BinaryExpr>(new AST::BinaryExpr(op,
+                                                                    std::move(expr),
+                                                                    std::move(right)));
+    }
+    return expr;
+}
+
+std::unique_ptr<AST::Expr> Parser::parseConcatExpr() {
+    std::unique_ptr<AST::Expr> expr = parseAddExpr();
+    while (m_currToken.type == TokenType::Ampersand) {
+        TokenType op = m_currToken.type;
+        eatToken(m_currToken.type);
+        auto right = parseAddExpr();
+        expr = std::unique_ptr<AST::BinaryExpr>(new AST::BinaryExpr(op,
+                                                                    std::move(expr),
+                                                                    std::move(right)));
+    }
+    return expr;
+}
+
+std::unique_ptr<AST::Expr> Parser::parseAddExpr() {
+    std::unique_ptr<AST::Expr> expr = parseMulExpr();
+    while (m_currToken.type == TokenType::Plus || m_currToken.type == TokenType::Minus) {
+        TokenType op = m_currToken.type;
+        eatToken(m_currToken.type);
+        std::unique_ptr<AST::Expr> right = parseMulExpr();
+        expr = std::unique_ptr<AST::BinaryExpr>(new AST::BinaryExpr(op,
+                                                                    std::move(expr),
+                                                                    std::move(right)));
+    }
+    return expr;
+}
+
+std::unique_ptr<AST::Expr> Parser::parseMulExpr() {
+    auto expr = parseUnaryExpr();
+    while (m_currToken.type == TokenType::Plus || m_currToken.type == TokenType::Minus) {
+        TokenType op = m_currToken.type;
+        eatToken(m_currToken.type);
+        auto right = parseUnaryExpr();
+        expr = std::unique_ptr<AST::BinaryExpr>(new AST::BinaryExpr(op,
+                                                                    std::move(expr),
```

Comment:
Why do we `std::move` here when there's an `std::move` in the constructor?

```cpp
    explicit BinaryExpr(TokenType op, std::unique_ptr<Expr> left, std::unique_ptr<Expr> right)
        : m_op(op)
        , m_left(std::move(left))
        , m_right(std::move(right)) { }
```

Not saying this is wrong, just wondering how this works

Reply:
I can check when I get home. You might be right. But if it is necessary, it’s probably because you need to pass the unique_ptr into the constructor function, and then the constructor function passes the unique_pointer into the newly constructed Expr.

Reply:
I've double checked and it is necessary, it won't compile unless it has both. So, yeah, I think it's because you first have to transfer ownership to the constructor function, which then transfers ownership to the newly-created object.